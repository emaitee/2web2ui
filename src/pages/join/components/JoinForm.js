/* eslint-disable */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, getFormValues } from 'redux-form';
import Recaptcha from 'react-recaptcha';

import config from 'src/config';
import { LINKS } from 'src/constants';
import { TextFieldWrapper, CheckboxWrapper } from 'src/components/reduxFormWrappers';
import { Button, UnstyledLink, Grid, Label } from '@sparkpost/matchbox';
import { required, minLength, email } from 'src/helpers/validation';
import styles from '../JoinPage.module.scss';
const { recaptcha } = config;

export class JoinForm extends Component {
  state = {
    reCaptchaReady: false,
    reCaptchaInstance: null
  }

  reCaptchaLoaded = () => {
    this.setState({ reCaptchaReady: true });
  }

  recaptchaVerifyCallback = (response) => {
    const { formValues, onSubmit } = this.props;
    this.reCaptchaInstance.reset();
    onSubmit({ ...formValues, recaptcha_response: response, recaptcha_type: 'invisible' });
  }

  executeRecaptcha = () => {
    this.reCaptchaInstance.execute();
  }

  render() {
    const {
      loading,
      pristine,
      invalid
    } = this.props;

    const { reCaptchaReady } = this.state;

    return (
      <form>
        <Grid className={styles.spacer}>
          <Grid.Column xs={12} md={6} lg={6}>
            <Field
              name='first_name'
              component={TextFieldWrapper}
              label='First Name'
              validate={required}
              disabled={!reCaptchaReady || loading}
            />
          </Grid.Column>
          <Grid.Column xs={12} md={6} lg={6}>
            <Field
              name='last_name'
              component={TextFieldWrapper}
              label='Last Name'
              validate={required}
              disabled={!reCaptchaReady || loading}
            />
          </Grid.Column>
        </Grid>
        <Field
          name='email'
          component={TextFieldWrapper}
          label='Email'
          validate={[required, email]}
          disabled={!reCaptchaReady || loading}
          autoComplete='username email'
        />
        <Field
          name='password'
          component={TextFieldWrapper}
          label='Password'
          validate={[required, minLength(8)]}
          disabled={!reCaptchaReady || loading}
          type='password'
          autoComplete='new-password'
        />
        <div className={styles.inline}>
          <Field
            name="tou_accepted"
            id='tou_accepted'
            component={CheckboxWrapper}
            type="checkbox"
            disabled={!reCaptchaReady || loading}
            validate={required}
          />
          <div className={styles.touLabel}>
            <Label id='tou_accepted'>
              I agree to SparkPost's <UnstyledLink to={LINKS.TOU} external>Terms of Use</UnstyledLink>
            </Label>
          </div>
        </div>

        <Button primary 
        disabled={loading || pristine || invalid || !reCaptchaReady }
        onClick={this.executeRecaptcha}
        >
          { loading ? 'Loading' : 'Create Account' }
        </Button>

        <Recaptcha
          ref={(e) => this.reCaptchaInstance = e}
          sitekey={recaptcha.invisibleKey}
          size="invisible"
          render='explicit'
          onloadCallback={this.reCaptchaLoaded}
          verifyCallback={this.recaptchaVerifyCallback}
        />

      </form>
    );
  }
}
const formName = 'registerAccountForm';
const mapStateToProps = (state, props) => ({
  initialValues: {
    tou_accepted: false
  },
  loading: state.account.createLoading,
  formValues: getFormValues(formName)(state)
});

const RegisterAccountForm = reduxForm({ form: formName })(JoinForm);
export default connect(mapStateToProps, {})(RegisterAccountForm);
