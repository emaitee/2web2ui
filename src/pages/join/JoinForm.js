import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field } from 'redux-form';

import { LINKS } from 'src/constants';
import { TextFieldWrapper, CheckboxWrapper } from 'src/components/reduxFormWrappers';
import { Button, UnstyledLink, Grid, Label } from '@sparkpost/matchbox';
import { required, minLength, email } from 'src/helpers/validation';
import styles from './JoinPage.module.scss';

export class JoinForm extends Component {

  render() {
    const {
      handleSubmit,
      submitting,
      pristine,
      invalid
    } = this.props;

    return (
      <form onSubmit={handleSubmit}>
        <Grid className={styles.spacer}>
          <Grid.Column xs={12} md={6} lg={6}>
            <Field
              name='first_name'
              component={TextFieldWrapper}
              label='First Name'
              validate={required}
              disabled={false}
            />
          </Grid.Column>
          <Grid.Column xs={12} md={6} lg={6}>
            <Field
              name='last_name'
              component={TextFieldWrapper}
              label='Last Name'
              validate={required}
              disabled={false}
            />
          </Grid.Column>
        </Grid>
        <Field
          name='email'
          component={TextFieldWrapper}
          label='Email'
          validate={[required, email]}
          disabled={false}
        />
        <Field
          name='password'
          component={TextFieldWrapper}
          label='Password'
          validate={[required, minLength(8)]}
          disabled={false}
          type='password'
          autoComplete='new-password'
          // data-lpignore={true} // removes inaccurate LastPass password management behavior
        />
        <div className={styles.inline}>
          <Field
            name='tou_accepted'
            id='tou_accepted'
            label=''
            validate={required}
            component={CheckboxWrapper}
          />
          <div className={styles.touLabel}>
            <Label id='tou_accepted'>
              I agree to SparkPost's <UnstyledLink to={LINKS.TOU} external>Terms of Use</UnstyledLink>
            </Label>
          </div>
        </div>

        <Button primary submit disabled={submitting || pristine || invalid }>{ submitting ? 'Loading' : 'Create Account' }</Button>
      </form>
    );
  }
}

const mapStateToProps = (state, props) => ({
  initialValues: { }
});

const RegisterUserReduxForm = reduxForm({ form: 'RegisterUser' })(JoinForm);
export default connect(mapStateToProps, {})(RegisterUserReduxForm);
