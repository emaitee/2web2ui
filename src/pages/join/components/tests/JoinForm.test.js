/* eslint max-lines: ["error", 200] */

import { shallow } from 'enzyme';
import React from 'react';
import { JoinForm } from '../JoinForm';

let props;
let instance;
let wrapper;


describe('JoinForm', () => {
  let mockRecaptcha;

  beforeEach(() => {
    mockRecaptcha = {
      execute: jest.fn(),
      reset: jest.fn()
    };

    props = {
      loading: false,
      onSubmit: jest.fn()
    };

    wrapper = shallow(<JoinForm {...props} />);
    instance = wrapper.instance();
    wrapper.setState({ reCaptchaInstance: mockRecaptcha });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('render', () => {
    it('disables form when loading', () => {
      wrapper.setProps({ loading: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('disables form when recaptcha is not loaded', () => {
      wrapper.setState({ reCaptchaReady: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('enables form correctly', () => {
      wrapper.setProps({ loading: false, pristine: false, invalid: false });
      wrapper.setState({ reCaptchaReady: true });
      expect(wrapper).toMatchSnapshot();
    });

    describe('submit', () => {
      it('invokes recaptcha challenge on click', () => {
        wrapper.find('Button').simulate('click');
        expect(mockRecaptcha.execute).toHaveBeenCalledTimes(1);
      });
    });

    describe('executeRecaptcha', () => {
      it('invokes recaptcha on click of submit button', () => {
        wrapper.setState({ reCaptchaInstance: mockRecaptcha });
        instance.executeRecaptcha();
        expect(mockRecaptcha.execute).toHaveBeenCalledTimes(1);
      });
    });

    describe('reCaptchaLoaded', () => {
      it('sets reCaptchaReady to be true', () => {
        instance.reCaptchaLoaded();
        expect(wrapper.state().reCaptchaReady).toBe(true);
      });
    });

    describe('linkRecaptcha', () => {
      it('sets reCaptchaInstance to state if not set', () => {
        instance.setState({ reCaptchaInstance: null });
        instance.linkRecaptcha({});
        expect(wrapper.state().reCaptchaInstance).toEqual({});
      });

      it('does not set reCaptchaInstance to state if already set', () => {
        instance.linkRecaptcha({});
        expect(wrapper.state().reCaptchaInstance).toEqual(mockRecaptcha);
      });
    });

    describe('recaptchaVerifyCallback', () => {
      it('resets recaptcha', () => {
        instance.recaptchaVerifyCallback('foobar');
        expect(mockRecaptcha.reset).toHaveBeenCalledTimes(1);
      });

      it('submits with correct data', () => {
        const formValues = { first_name: 'foo', last_name: 'bar', email: 'foo@bar.com', password: '***' };
        wrapper.setProps({ formValues });
        instance.recaptchaVerifyCallback('foobar');
        expect(props.onSubmit).toHaveBeenCalledWith({ ...formValues, recaptcha_response: 'foobar', recaptcha_type: 'invisible' });
      });
    });

  });
});
