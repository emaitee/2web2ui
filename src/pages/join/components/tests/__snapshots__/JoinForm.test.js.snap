// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinForm render disables form when loading 1`] = `
<form>
  <Grid>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={true}
        label="First Name"
        name="first_name"
        validate={[Function]}
      />
    </Grid.Column>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={true}
        label="Last Name"
        name="last_name"
        validate={[Function]}
      />
    </Grid.Column>
  </Grid>
  <Field
    autoComplete="username email"
    component={[Function]}
    disabled={true}
    label="Email"
    name="email"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    autoComplete="new-password"
    component={[Function]}
    disabled={true}
    label="Password"
    name="password"
    type="password"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <div>
    <Field
      component={[Function]}
      disabled={true}
      id="tou_accepted"
      name="tou_accepted"
      type="checkbox"
      validate={[Function]}
    />
    <div>
      <Label
        id="tou_accepted"
      >
        I agree to SparkPost's 
        <UnstyledLink
          external={true}
          to="https://www.sparkpost.com/policies/tou"
        >
          Terms of Use
        </UnstyledLink>
      </Label>
    </div>
  </div>
  <Button
    disabled={true}
    onClick={[Function]}
    primary={true}
    size="default"
  >
    Loading
  </Button>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</form>
`;

exports[`JoinForm render disables form when recaptcha is not loaded 1`] = `
<form>
  <Grid>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={true}
        label="First Name"
        name="first_name"
        validate={[Function]}
      />
    </Grid.Column>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={true}
        label="Last Name"
        name="last_name"
        validate={[Function]}
      />
    </Grid.Column>
  </Grid>
  <Field
    autoComplete="username email"
    component={[Function]}
    disabled={true}
    label="Email"
    name="email"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    autoComplete="new-password"
    component={[Function]}
    disabled={true}
    label="Password"
    name="password"
    type="password"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <div>
    <Field
      component={[Function]}
      disabled={true}
      id="tou_accepted"
      name="tou_accepted"
      type="checkbox"
      validate={[Function]}
    />
    <div>
      <Label
        id="tou_accepted"
      >
        I agree to SparkPost's 
        <UnstyledLink
          external={true}
          to="https://www.sparkpost.com/policies/tou"
        >
          Terms of Use
        </UnstyledLink>
      </Label>
    </div>
  </div>
  <Button
    disabled={true}
    onClick={[Function]}
    primary={true}
    size="default"
  >
    Create Account
  </Button>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</form>
`;

exports[`JoinForm render enables form correctly 1`] = `
<form>
  <Grid>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={false}
        label="First Name"
        name="first_name"
        validate={[Function]}
      />
    </Grid.Column>
    <Grid.Column
      lg={6}
      md={6}
      xs={12}
    >
      <Field
        component={[Function]}
        disabled={false}
        label="Last Name"
        name="last_name"
        validate={[Function]}
      />
    </Grid.Column>
  </Grid>
  <Field
    autoComplete="username email"
    component={[Function]}
    disabled={false}
    label="Email"
    name="email"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    autoComplete="new-password"
    component={[Function]}
    disabled={false}
    label="Password"
    name="password"
    type="password"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <div>
    <Field
      component={[Function]}
      disabled={false}
      id="tou_accepted"
      name="tou_accepted"
      type="checkbox"
      validate={[Function]}
    />
    <div>
      <Label
        id="tou_accepted"
      >
        I agree to SparkPost's 
        <UnstyledLink
          external={true}
          to="https://www.sparkpost.com/policies/tou"
        >
          Terms of Use
        </UnstyledLink>
      </Label>
    </div>
  </div>
  <Button
    disabled={false}
    onClick={[Function]}
    primary={true}
    size="default"
  >
    Create Account
  </Button>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</form>
`;
